cep-recipe-serialization.md

```diff
@@ -0,0 +1,1110 @@
+# Recipe format: Rendered recipe YAML and serialization
+
+## Motivation
+
+We want to specify everything that goes into the final package in a standardized
+format. This defines the structure and contents of the rendered_recipe.yaml file
+which captures the complete recipe, build configuration, dependencies and
+sources used to create a conda package.
+
+The goal is to have a single file that contains all the information needed to
+reproduce an exact build of a package. This will enable better reproducibility,
+provenance tracking, and debugging of conda packages.
+
+## Storage
+
+The rendered recipe YAML file is stored in the `info/recipe` folder of the final
+package. Alongside the rendered recipe, we copy the entire "recipe folder".
+If the original recipe was not named `recipe.yaml`, then we rename it inside the package.
+
+For example:
+
+```txt
+├── recipe
+    ├── patches
+    │   └── patch1.diff
+    ├── build.sh
+    ├── build.bat
+    └── my_recipe.yaml
+```
+
+Will be copied into the following folder structure in the final package:
+
+```txt
+├── ...
+├── index.json
+└── recipe
+    ├── build.bat
+    ├── build.sh
+    ├── recipe.yaml
+    ├── patches
+    │   └── patch1.diff
+    ├── rendered_recipe.yaml
+    └── variant_config.yaml
+```
+
+## Specification
+
+The rendered recipe YAML file contains the following top-level sections:
+
+1. `recipe`: The fully rendered recipe with all Jinja2 expressions evaluated and
+   if/else logic resolved to the final used values.
+2. `build_configuration`: Captures all the build-time configuration such as the
+   target platform, build/host prefixes, variant used, package version and build
+   string, channels, etc.
+3. `finalized_dependencies`: Records the exact versions and builds of all
+   dependencies (build, host and run) that were used at build time, like a
+   "lockfile".
+4. `finalized_sources`: Contains the final references to the sources used, e.g.
+   git commit hashes, tarball URLs + sha256, etc.
+5. `system_tools`: Records the tools that were used from the host system in
+   order to warn if reproducing the build on a different host system might lead
+   to different results (patchelf, git, etc).
+
+### Recipe section
+
+This section is the fully rendered recipe in valid YAML format, with all Jinja2
+expressions evaluated and if/else logic resolved to the final values used during
+the build.
+
+Some key things to note:
+
+- Compiler expressions like ${{ compiler('c') }} are rendered to their final
+  YAML dictionary form:
+
+  ```yaml
+  - ${{ compiler('c') }}
+  # becomes
+  - compiler: c
```

Comment:
We currently have an "intermediate" form in rattler-build.  We need that for `pin_subpackage` because that needs to have the rest of the recipe rendered first. 
However, for `compiler` we can make the function return the final value instead. It would have some other benefits (e.g. that we can use the `compiler` function in other contexts easily).